### ISSUE TITLE
Does WhisperState::get_logits(segment: c_int) need the segment parameter?

### ISSUE THREAD

**Issue by jcsoo** [tazz4843/whisper-rs#49]
My understanding is that n_tokens should always be equal to the number of tokens in the last segment. In that case, get_logits would not return the correct number of rows (except by accident) if called with any other segment index.

In that case, should we remove the `segment` parameter and use `full_n_segments()` to get the segment index of the last segment instead?

```
    pub fn get_logits(&self, segment: c_int) -> Result<Vec<Vec<f32>>, WhisperError> {
        let ret = unsafe { whisper_rs_sys::whisper_get_logits_from_state(self.ptr) };
        if ret.is_null() {
            return Err(WhisperError::NullPointer);
        }
        let mut logits = Vec::new();
        let n_vocab = self.n_vocab();
        let n_tokens = self.full_n_tokens(segment)?;
        for i in 0..n_tokens {
            let mut row = Vec::new();
            for j in 0..n_vocab {
                let idx = (i * n_vocab) + j;
                let val = unsafe { *ret.offset(idx as isize) };
                row.push(val);
            }
            logits.push(row);
        }
        Ok(logits)
    }
```

Additionally, the implementation is allocating a Vec of Vec<f32> on every call. It would be great to have an additional function `WhisperState::get_logits_raw() -> &[f32] {}` that would provide direct read-only access to the underlying logits array slice. This could then be passed to a higher level wrapper which could provide access to individual rows, or used with an external 2D matrix library such as `ndarray` or `nalgebra`, maybe as an optional dependency.

I can start a pull request for this if it seems like it would be helpful.

**Comment by jcsoo** [comment]
By the way, please don't take these requests as criticism. This crate has been incredibly useful to me already, and I'd like to contribute back as much as possible. Thank you for all the great work so far!

**Comment by tazz4843** [comment]
> My understanding is that n_tokens should always be equal to the number of tokens in the last segment.
I believe so, however at the time, looking through the whisper.cpp source, I couldn't confirm this. If you can confirm it, feel free to do a PR.

> Additionally, the implementation is allocating a Vec of Vec on every call. It would be great to have an additional function `WhisperState::get_logits_raw() -> &[f32] {}` that would provide direct read-only access to the underlying logits array slice.
Not sure why I never implemented this myself, seems rather easy. PR for this as well would be appreciated.

**Comment by jcsoo** [comment]
In fact, looking at [whisper_decode_internal](https://github.com/ggerganov/whisper.cpp/blob/14bee39b290f81381a259c53a169ffac8fcce987/whisper.cpp#L2261), the current implementation only copies logits for the last token into whisper_state.

```
    // extract logits for all N tokens
    //logits_out.resize(N*n_vocab);
    //memcpy(logits_out.data(), ggml_get_data(logits), sizeof(float)*N*n_vocab);

    // extract logits only for the last token
    logits_out.resize(n_vocab);
    memcpy(logits_out.data(), ggml_get_data(logits), sizeof(float)*n_vocab);
```

Looking through the rest of whisper.cpp, there isn't any other place that appends or modifies the logits in whisper_state. I think this is a case of the comments in the structure not matching up with current usage.

So maybe `get_logits()` should not take any segment argument at all and simply return a f32 slice with length n_vocab. Or, if you want to support the future case where there may be more than one set of logits, `get_logits(index: usize) -> &[f32]` and panic if index > 0.

**Comment by tazz4843** [comment]
Interesting. I think the latter would be a better choice, but perhaps returning an error would be more user-friendly over panicking.

**Comment by tazz4843** [comment]
Closed by #52
